controllers/apps/transformer_component_workload.go

```diff
@@ -657,41 +739,95 @@ func (r *componentWorkloadOps) leaveMember4ScaleIn() error {
 		return err
 	}
 
-	// TODO: Move memberLeave to the ITS controller. Instead of performing a switchover, we can directly scale down the non-leader nodes. This is because the pod ordinal is not guaranteed to be continuous.
-	podsToMemberLeave := make([]*corev1.Pod, 0)
-	for _, pod := range pods {
-		// if the pod not exists in the generated pod names, it should be a member that needs to leave
-		if _, ok := r.desiredCompPodNameSet[pod.Name]; ok {
-			continue
+	if !(isLeader(pod) || // if the pod is leader, it needs to call switchover
+		(r.synthesizeComp.LifecycleActions != nil && r.synthesizeComp.LifecycleActions.MemberLeave != nil)) { // if the memberLeave action is defined, it needs to call it
+		return nil
+	}
+
+	lfa, err := lifecycle.New(r.synthesizeComp, pod, pods...)
+	if err != nil {
+		return err
+	}
+
+	// switchover if the leaving pod is leader
+	if switchoverErr := tryToSwitchover(lfa, pod); switchoverErr != nil {
+		return switchoverErr
+	}
+
+	if err = lfa.MemberLeave(r.reqCtx.Ctx, r.cli, nil); err != nil {
+		if !errors.Is(err, lifecycle.ErrActionNotDefined) && err == nil {
+			return err
 		}
-		podsToMemberLeave = append(podsToMemberLeave, pod)
 	}
-	for _, pod := range podsToMemberLeave {
-		if !(isLeader(pod) || // if the pod is leader, it needs to call switchover
-			(r.synthesizeComp.LifecycleActions != nil && r.synthesizeComp.LifecycleActions.MemberLeave != nil)) { // if the memberLeave action is defined, it needs to call it
+	return nil
+}
+
+func (r *componentWorkloadOps) checkAndDoMemberJoin() error {
+	podsToMemberjoin := getPodsToMemberJoinFromAnno(r.runningITS)
+
+	if r.synthesizeComp.LifecycleActions == nil || r.synthesizeComp.LifecycleActions.MemberJoin == nil {
+		podsToMemberjoin.Clear()
+	}
+	err := r.doMemberJoin(podsToMemberjoin)
+	if err != nil {
+		return err
+	}
+
+	if podsToMemberjoin.Len() == 0 {
+		delete(r.protoITS.Annotations, constant.MemberJoinStatusAnnotationKey)
```

Comment:
Would this conflict with the operation of constant.MemberJoinStatusAnnotationKey in annotateInstanceSetForMemberJoin? Consider a scenario where getPodsToMemberJoinFromAnno(r.runningITS) returns empty because the annotation hasn't been updated in annotateInstanceSetForMemberJoin yet?

Reply:
added check for protoITS's membeJoin Anno, if not blank, do nothing and wait for next Reconcile